### PR TITLE
Optimized inference of GPT-NEO model on HPU

### PR DESCRIPTION
**Pull request by XinyuYe-Intel**
# What does this PR do?

Optimized inference of GPT-NEO model on HPU.

## Before submitting
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: docs/source/index.mdx [hunk 1]
```diff
@@ -45,6 +45,7 @@ In the tables below, ✅ means single-card, multi-card and DeepSpeed have all be
 | BLOOM(Z)     |          | <div style="text-align:left"><li>DeepSpeed</li></div> | <li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
 | StarCoder / StarCoder2 | ✅ | <div style="text-align:left"><li>Single card</li></div> | <li>[language modeling](https://github.com/huggingface/optimum-habana/tree/main/examples/language-modeling)</li><li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
 | GPT-J        | <div style="text-align:left"><li>DeepSpeed</li></div> | <div style="text-align:left"><li>Single card</li><li>DeepSpeed</li></div> | <li>[language modeling](https://github.com/huggingface/optimum-habana/tree/main/examples/language-modeling)</li><li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
+| GPT-Neo      |          | <div style="text-align:left"><li>Single card</li></div> | <li>[language modeling](https://github.com/huggingface/optimum-habana/tree/main/examples/language-modeling)</li><li>[text generation](https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation)</li> |
```
Comment (regisss) [inline docs/source/index.mdx:48]: Can you add the same in the README? https://github.com/huggingface/optimum-habana/blob/fa1fbc5f81ab1012263d36c4b73201db8ecd05e5/README.md?plain=1#L192

File: tests/test_text_generation_example.py [hunk 2]
```diff
@@ -42,6 +42,7 @@
             ("google/gemma-7b", 1, False, 109.70751574382221),
             ("state-spaces/mamba-130m-hf", 1536, False, 8600),
             ("Deci/DeciLM-7B", 1, False, 120),
+            ("EleutherAI/gpt-neo-2.7B", 1, False, 223.52655640343593),
```
Comment (regisss) [inline tests/test_text_generation_example.py:45]: Can you add a baseline for Gaudi1 too?

File: docs/source/index.mdx (hunk 1 above)
Comment (XinyuYe-Intel) [reply docs/source/index.mdx:48]: updated README.

File: tests/test_text_generation_example.py (hunk 2 above)
Comment (XinyuYe-Intel) [reply tests/test_text_generation_example.py:45]: We don't have the Gaudi1 machine.

**Comment by jiminha** [comment]
@XinyuYe-Intel  
 Can you provide your full command for the inference and provide the perf/accuracy value for both before/after your patch?

File: optimum/habana/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+    return BaseModelOutputWithPast(
+        last_hidden_state=hidden_states,
+        past_key_values=presents,
+        hidden_states=all_hidden_states,
+        attentions=all_self_attentions,
+    )
+
+
+class GaudiGPTNeoForCausalLM(GPTNeoForCausalLM):
+    """
```
Comment (jiminha) [inline optimum/habana/transformers/models/gpt_neo/modeling_gpt_neo.py:299]: Any reason why this has class/forward function instead of just forward function mapped?

File: optimum/habana/transformers/models/gpt_neo/modeling_gpt_neo.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+class GaudiGPTNeoBlock(GPTNeoBlock):
+    def __init__(self, config, layer_id):
+        super(GPTNeoBlock, self).__init__()
+        hidden_size = config.hidden_size
+        inner_dim = config.intermediate_size if config.intermediate_size is not None else 4 * hidden_size
+        self.ln_1 = torch.nn.LayerNorm(hidden_size, eps=config.layer_norm_epsilon)
+        self.attn = GPTNeoAttention(config, layer_id)
+        self.ln_2 = torch.nn.LayerNorm(hidden_size, eps=config.layer_norm_epsilon)
+        self.mlp = GPTNeoMLP(inner_dim, config)
+
```
Comment (jiminha) [inline optimum/habana/transformers/models/gpt_neo/modeling_gpt_neo.py:111]: Any changes in init function here from the original one? If not, just map to forward function only as well.

File: optimum/habana/transformers/models/gpt_neo/modeling_gpt_neo.py (hunk 3 above)
Comment (XinyuYe-Intel) [reply optimum/habana/transformers/models/gpt_neo/modeling_gpt_neo.py:286]: Since `prepare_inputs_for_generation` of `GaudiGPTNeoForCausalLM` is also changed.

File: optimum/habana/transformers/models/gpt_neo/modeling_gpt_neo.py (hunk 4 above)
Comment (XinyuYe-Intel) [reply optimum/habana/transformers/models/gpt_neo/modeling_gpt_neo.py]: Changed to forward function only.

**Comment by XinyuYe-Intel** [comment]
> @XinyuYe-Intel Can you provide your full command for the inference and provide the perf/accuracy value for both before/after your patch?

Below is the command I used for the inference.
```
python run_generation.py --model_name_or_path EleutherAI/gpt-neo-2.7B --batch_size 1 --use_hpu_graphs  --bf16  --do_sample --use_kv_cache
```
Below is the perf before optimization:
image
and the perf after optimization:
image

File: docs/source/index.mdx (hunk 1 above)
Comment (regisss) [inline docs/source/index.mdx:48]: Here there are "language modeling" and "text generation" but in the README there is only "text generation". Since the focus of this PR is inference, I guess it should be only "text generation" right?

Comment (XinyuYe-Intel) [reply docs/source/index.mdx]: Yes, you are right, I've changed it to text generation only, thanks for your reminding.

**Review by regisss — APPROVED** [review]
